### DISCUSSION TITLE
No eGPU attached display detected with open source drivers

### DISCUSSION THREAD

**Discussion by dannycolin** [hertg/egpu-switcher#82] [Q&A] [answered]
I got the following issue when trying to switch to my AMD eGPU.

`No eGPU attached display detected with open source drivers. (Of 2 eGPU outputs detected) Internal mode and setting DRI_PRIME variable are recommended for this configuration.`

I can still boot the Xorg session but there's a lot of lag.

Any idea @hertg?

**Comment by hertg** [comment]
Do you not have any monitors connected to the eGPU directly?
There is the `--override` flag on the `egpu-switcher switch` command, did you try that?

```
If an AMD GPU or open-source NVIDIA drivers are used, the "switch" command 
will prevent from switching to the eGPU if there are no displays directly attached to it. 
This flag will make sure to switch to the EGPU even if there are no displays attached.
```

Unfortunately, I do not have an AMD GPU that fits into my eGPU enclosure, so I am not too familiar with this part of the script as it also was an external contribution in #14. Maybe check out the comments on the PR to get a better understanding what's going on.

**Comment by dannycolin** [comment]
> Do you not have any monitors connected to the eGPU directly?

You don't need to plug an external monitor on a eGPU if it's hooked up to a laptop. It
should use the internal monitor in this case. That's exactly how my setup with my NVDIA
card was working before switching to the AMD one.

The strange thing is the AMDGPU driver are loaded but it's very laggy and there doesn't seem to be a lot of workload on the AMD GPU. I'm guessing for some reason the GPU and the internal monitor aren't communicating properly and the graphics must be offloaded to the CPU internal GPU resulting in poor performance (lag).

In any case, there's something going wrong there. Do you have more information on why the script can't swithc to the eGPU if there's no external monitor with the AMDGPU driver? That seems an odd requirement for a eGPU since the whole concept is to add a GPU to a laptop and that it works with an NVIDIA + proprietary driver or on Windows.

Thanks for the link to the PR. I'll have a look at it next week when I get some free time :).

**Comment by hertg** [comment]
No unfortunately, as I've said, I do not really have more knowledge about how it works, and why it behaves like that. Maybe @ewagner12 might have more information, as they are the code owner of that specific piece of code.

**Comment by ewagner12** [comment, marked as answer]
I added this check and code to the script specifically because of the lag that you mentioned. The open source drivers (both amdgpu and nouveau) have this issue of lag when you try to switch to the eGPU without a display attached. The suggested way to use the eGPU instead is to add the DRI_PRIME=1 environment variable before any command or game you want to use the eGPU. In reality, manually adding the variable shouldn't even be necessary sometimes since applications like Steam and Lutris will pickup and use the eGPU by default.

But if we wanted to add a way of switching to an AMD eGPU without this lag issue it might be possible to use the amdgpu.virtual_display module parameter to spoof a virtual display. [https://docs.kernel.org/gpu/amdgpu/module-parameters.html]( https://docs.kernel.org/gpu/amdgpu/module-parameters.html)

**Reply by dannycolin** [reply]
Is there a way to setDRI_PRIME globally? thanks a lot for that valuable information.

**Reply by ewagner12** [reply]
I've had luck with putting DRI_PRIME=1 in /etc/environment

**Reply by dannycolin** [reply]
Quick follow-up question, do you know if this driver limitation also affect Wayland?

**Reply by ewagner12** [reply]
It is not an issue I've had in any of the wayland desktops I have used (Sway, GNOME, and KDE Plasma).
